File: luogu/P1213-bfs.py
###
# Solution: BFS + State Compress, TLE with python code
# each clock has 4 status: 1 (3:00), 2 (6:00), 3 (9:00), 0 (0:00)
# each step make some clocks s to be (s+1)%4, final goal is to be 000000000
# We use BFS to search with both the step num and the strategy in increasing order
###

import collections
final_status = '000000000'

actions = [[0, 1, 3, 4], [0, 1, 2], [1, 2, 4, 5], [0, 3, 6],
           [1, 3, 4, 5, 7], [2, 5, 8], [3, 4, 6, 7],
           [6, 7, 8], [4, 5, 7, 8]]

status = []
for i in range(3):
    arr = input().split()
    arr = [int(_)  for _ in arr]
    for num in arr:
        status.append((num//3)%4)

status_steps = dict()
s0 = ''.join([str(_) for _ in status])
status_steps[s0] = [0, 0, ''] # step num, action type, pred status
if s0 == final_status:
    print('')
else:
    dq = collections.deque()
    dq.append(s0)
    while True:
        if len(dq) == 0:
            break
        st = dq.popleft()
        step_num, _, _ = status_steps[st]
        for i, step in enumerate(actions):
            stat = [int(char) for char in st]
            for idx in step:
                stat[idx] = (stat[idx] + 1) % 4
            snew = ''.join([str(_) for _ in stat])
            if snew not in status_steps:
                status_steps[snew] = [step_num + 1, i+1, st]
                dq.append(snew)
        if final_status in status_steps:
            cur_status = final_status
            pathes = []
            while True:
                _, type, cur_status = status_steps[cur_status]
                if cur_status == '':
                    break
                pathes.append(type)

            print(' '.join([str(_) for _ in pathes[::-1]]))
            break

File: luogu/P1213-bfs.cpp
/*
Solution can be found in P1213.py
*/

#include <iostream>
#include <queue>
using namespace std;
struct STATUS{
    int step_num, action_type, pred_status;
    bool has_visit;
    STATUS() {
        has_visit = false;
    }
    void set_value(int sn, int at, int ps) {
        this->step_num = sn, this->action_type = at, this->pred_status = ps;
        has_visit = true;
    }
};
int actions[9][5] = {{0, 1, 3, 4, -1}, {0, 1, 2, -1, -1}, {1, 2, 4, 5, -1}, {0, 3, 6, -1, -1}, {1, 3, 4, 5, 7}, {2, 5, 8, -1, -1}, {3, 4, 6, 7, -1}, {6, 7, 8, -1, -1}, {4, 5, 7, 8, -1}};
int status[9];
const int NMAX = 1 << 18;
STATUS status_data[NMAX];
int status2num() {
    int _sum = 0;
    for (int i = 0; i < 9; i++) {
        _sum = _sum * 4 + status[i];
    }
    return _sum;
}
void num2status(int num) {
    for (int i = 8; i >= 0; i--) {
        status[i] = num % 4;
        num /= 4;
    }
}
int main() {
    int final_status = 0;
    int a, b, c;
    for (int i = 0; i < 3; i++) {
        scanf("%d%d%d", &a, &b, &c);
        status[i*3] = (a/3)%4;
        status[i*3+1] = (b/3)%4;
        status[i*3+2] = (c/3)%4;
    }
    int s0 = status2num();
    if (s0 == final_status) {
        return 0;
    }
    queue<int> dq;
    dq.push(s0);
    status_data[s0].set_value(0, -1, -1);
    while (true) {
        if (dq.size() == 0) { break; }
        int st = dq.front(); dq.pop();
        for (int i = 0; i < 9; i++) {
            num2status(st);
            for (int j = 0; j < 5; j++) {
                int idx = actions[i][j];
                if (idx > -1) {
                    status[idx] = (status[idx] + 1) % 4;
                }
            }
            int s_new = status2num();
            if (!status_data[s_new].has_visit) {
                status_data[s_new].set_value(status_data[st].step_num+1, i + 1, st);
                dq.push(s_new);
                // printf("visit %d from %d in type %d\n", s_new, st, i+1);
            }
        }
        if (status_data[final_status].has_visit) {
            int cur_state = final_status;
            vector<int> pathes;
            while (true) {
                if (status_data[cur_state].pred_status == -1) {
                    break;
                }
                pathes.push_back(status_data[cur_state].action_type);
                cur_state = status_data[cur_state].pred_status;
            }
            for (int i = pathes.size()-1; i >= 0; i--) {
                printf("%d ", pathes[i]);
            }
            printf("\n");
            break;
        }
        


    }


    return 0;
}